xmppserver/src/main/java/org/jivesoftware/util/cache/ExternalizableUtil.java

```diff
@@ -35,14 +41,37 @@
  */
 public class ExternalizableUtil {
 
-    private static ExternalizableUtil instance = new ExternalizableUtil();
+    private static ExternalizableUtil instance;
 
     private ExternalizableUtilStrategy strategy = new DefaultExternalizableUtil();
 
     static {
         instance = new ExternalizableUtil();
     }
 
+    /**
+     * Pool of SAX Readers. SAXReader is not thread safe so we need to have a pool of readers.
+     */
+    private final static int POOL_SIZE = 10;
+    private final BlockingQueue<SAXReader> xmlReaders = new LinkedBlockingQueue<>(POOL_SIZE);
+
+    private SAXReader takeSAXReader() throws InterruptedException {
+        SAXReader xmlReader;
+        if (! xmlReaders.isEmpty()) {
+            xmlReader = new SAXReader();
+            xmlReader.setEncoding("UTF-8");
+        } else {
+            xmlReader = xmlReaders.take();
+        }
+        return xmlReader;
+    }
```

Comment:
Is the `takeSAXReader` implementation correct? If the collection is _not_ empty, create and return a new reader. If the collection _is_ empty, block until it becomes non-empty?